packages/jellyfish-transaction-builder/__tests__/provider.mock.ts

```diff
@@ -0,0 +1,132 @@
+import { BigNumber } from 'bignumber.js'
+import { Bech32, EllipticPair, HASH160, WIF } from '@defichain/jellyfish-crypto'
+import { EllipticPairProvider, FeeRateProvider, Prevout, PrevoutProvider } from '../src'
+import { MasterNodeRegTestContainer } from '@defichain/testcontainers'
+import { OP_CODES, Script } from '@defichain/jellyfish-transaction'
+import { randomEllipticPair } from './test.utils'
+
+export class MockFeeRateProvider implements FeeRateProvider {
+  constructor (
+    public readonly container: MasterNodeRegTestContainer
+  ) {
+  }
+
+  async estimate (): Promise<BigNumber> {
+    const result = await this.container.call('estimatesmartfee', [1])
+    if (result.errors !== undefined && result.errors.length > 0) {
+      return new BigNumber('0.00005000')
+    }
+    return new BigNumber(result.feerate)
+  }
+}
+
+export class MockPrevoutProvider implements PrevoutProvider {
+  constructor (
+    public readonly container: MasterNodeRegTestContainer,
+    public ellipticPair: EllipticPair
+  ) {
+  }
+
+  async all (): Promise<Prevout[]> {
+    const pubKey = await this.ellipticPair.publicKey()
+    const unspent: any[] = await this.container.call('listunspent', [
+      1, 9999999, [Bech32.fromPubKey(pubKey, 'bcrt')]
+    ])
+
+    return unspent.map((utxo: any): Prevout => {
+      return MockPrevoutProvider.mapPrevout(utxo, pubKey)
+    })
+  }
+
+  async collect (minBalance: BigNumber): Promise<Prevout[]> {
```

Comment:
Is this intentional for the arg `minBalance` is unused? What's the expected behavior for non mock case? It should stop collecting once exceeded the required amount?

Reply:
> Collect the minimum number of Prevout required to create a transaction. As PrevoutProvider allows an agnostic implementation, it is free to use any UTXO selector patterns.

For mocking yes; and essentially it implements a collating pattern. "I think I left a reason below why I didn't use it, due to issues with defid.:

For non-mock case, there are a few patterns available for implementation, it's up to the interface implementor to decide between:

- **forward**: Selecting UTXO without sorting
- **crumbs**: Selecting UTXO by ascending the balance
- **pie**: Selecting UTXO by descending the balance
- **all**: Selecting all UTXO regardless of `minBalance`

Ultimately it's up to the implementor to decide since it depends on what indexes are available for the implementor.